modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java

```diff
@@ -0,0 +1,91 @@
+import org.apache.lucene.index.ReaderUtil;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.TopDocs;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.percolator.PercolatorHighlightSubFetchPhase.locatePercolatorQuery;
+
+/**
+ * Adds a special field to the a percolator query hit to indicate which documents matched with the percolator query.
+ * This is useful when multiple documents are being percolated in a single request.
+ */
+class PercolatorMatchedSlotSubFetchPhase implements FetchSubPhase {
+
+    static final String FIELD_NAME = "_percolator_document_slot";
+
+    @Override
+    public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOException {
+        PercolateQuery percolateQuery = locatePercolatorQuery(context.query());
+        if (percolateQuery == null) {
+            return;
+        }
+
+        if (percolateQuery.getDocuments().size() == 1) {
+            for (SearchHit hit : hits) {
+                Map<String, DocumentField> fields = hit.fieldsOrNull();
+                if (fields == null) {
+                    fields = new HashMap<>();
+                    hit.fields(fields);
+                }
+                fields.put(FIELD_NAME, new DocumentField(FIELD_NAME, Collections.singletonList(0)));
+            }
+            return;
+        }
+
+        PercolateQuery.QueryStore queryStore = percolateQuery.getQueryStore();
+        List<LeafReaderContext> ctxs = context.searcher().getIndexReader().leaves();
+        IndexSearcher percolatorIndexSearcher = percolateQuery.getPercolatorIndexSearcher();
+        for (SearchHit hit : hits) {
+            LeafReaderContext ctx = ctxs.get(ReaderUtil.subIndex(hit.docId(), ctxs));
+            int segmentDocId = hit.docId() - ctx.docBase;
+            Query query = queryStore.getQueries(ctx).apply(segmentDocId);
+
+            int n = percolatorIndexSearcher.getIndexReader().maxDoc();
+            TopDocs topDocs = percolatorIndexSearcher.search(query, n, new Sort(SortField.FIELD_DOC));
+            Map<String, DocumentField> fields = hit.fieldsOrNull();
+            if (fields == null) {
+                fields = new HashMap<>();
+                hit.fields(fields);
+            }
+            List<Object> slots = Arrays.stream(topDocs.scoreDocs)
+                .map(scoreDoc -> scoreDoc.doc)
+                .collect(Collectors.toList());
+            fields.put(FIELD_NAME, new DocumentField(FIELD_NAME, slots));
```

Comment:
I don't think reusing Lucene doc ids as slots will work in case of nested docs.

Reply:
ah good point. Let me think about this.